### ISSUE TITLE
on_internal_error should always log its message

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#955]
Our `on_internal_error()` implementation is:
```c++
void seastar::on_internal_error(logger& logger, std::exception_ptr ex) {
    if (abort_on_internal_error.load()) {
        logger.error("{}", ex);
        abort();
    } else {
        std::rethrow_exception(std::move(ex));
    }
}
```

We have two modes - `abort_on_internal_error` _true_ and _false_.  As we can see above, in the _true_ mode, we print a log message and abort. In the `false` case we throw an exception - **but do not log the message**. I think that this is a mistake, and we should log a message in this case too.

We saw an example of why such logging is needed in the Scylla project, in issue https://github.com/scylladb/scylla/issues/9422. In that case, some inner code detected an internal error and called `on_internal_error()` which caused it to throw an exception. The calling code was not prepared to handle exceptions, and the exception caused a second problem which caused it to throw **a different** exception. At that point the exception from `on_internal_error()` was lost and we didn't even know that it happened. Had we got the internal error message in the log, we would have known that it caused the cascade of errors.

One downside of logging the error message is that it might cause double logging - we log the message, and then some upper-level caller logs it again. However, I'm not sure this is a big problem - it might actually be useful to know both about the inner code that caught the internal error itself, and about the outer code that called it and finally logged it.

Another downside of logging is that in theory some bugs can cause us to go into an infinite loop of `abort_on_internal_error()` and spamming the log. But I suspect that if we have a loop of exceptions we anyway will have some caller reporting them and spamming the log, so it's not a new problem.

**Comment by psarna** [comment]
I agree, double logging is a negligible problem (or even not a problem at all), if it helps us pinpoint an internal bug. And `on_internal_error` generally indicates a bug, so it's never expected on any fast path. I'll be happy to queue such a fix when you post a patch.

**Comment by bhalevy** [comment]
Sounds good to me.
@denesb are you ok with this too?

**Comment by denesb** [comment]
> Sounds good to me. @denesb are you ok with this too?

Yes, sounds like a good idea.

**Comment by nyh** [comment]
@tgrabiec in the specific example code that prompted me to open this issue (https://github.com/scylladb/scylla/issues/9422), the second exception was another `on_internal_error()`, or more accurately `on_internal_error_noexcept()` to force an abort. The flow is as follows:

1. Object A is moved into object B's constructor.
2. Object B's constructor detects an internal error, and uses on_internal_error() which throws.
3. Since object A was moved, it is destroyed when object B's constructor exists with an exception.
4. Object A's destructor notices it is destroyed without some expected cleanup it requires, and uses on_internal_error_abort() to abort at this point.

We could modify `on_internal_error*()` to use a nested exception if there is a `std::current_exception()`, 
but I think that won't help: step 4 is **during** stack unwinding, so I think Object A's destructor, even if it wanted to figure out if there is a current exception, won't know it. I even have a hunch that if step 4 used `on_internal_error` (the one throwing an exception) we would get a `terminate()` because C++ can't handle such double exceptions.

In any case, `on_internal_error()` is obviously meant for debugging incorrect programs, not for perfectly written programs. So wouldn't it be useful that the error that was already detected doesn't get hidden even if there is a second bug and the program's exception safety isn't perfect?

**Comment by nyh** [comment]
Fixed by @kbr- in commit 84b042759046259c39022abcdefc56ee9110d524 from pull request https://github.com/scylladb/seastar/pull/1121
